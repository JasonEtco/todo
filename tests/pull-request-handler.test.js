const { gimmeApp, loadConfig, loadDiff } = require('./helpers')
const pullRequestOpened = require('./fixtures/payloads/pull_request.opened.json')

describe('pull-request-handler', () => {
  let app, github
  const event = { name: 'pull_request', payload: pullRequestOpened }

  beforeEach(() => {
    const gimme = gimmeApp()
    app = gimme.app
    github = gimme.github
  })

  it('comments on a pull request', async () => {
    await app.receive(event)
    expect(github.issues.createComment).toHaveBeenCalledTimes(1)
    expect(github.issues.createComment.mock.calls[0]).toMatchSnapshot()
  })

  it('comments on a pull request and mentions the assigned user', async () => {
    github.repos.getContent.mockReturnValueOnce(loadConfig('autoAssignString'))
    await app.receive(event)
    expect(github.issues.createComment.mock.calls[0]).toMatchSnapshot()
  })

  it('comments on a pull request and mentions the assigned users', async () => {
    github.repos.getContent.mockReturnValueOnce(loadConfig('autoAssignArr'))
    await app.receive(event)
    expect(github.issues.createComment.mock.calls[0]).toMatchSnapshot()
  })

  it('does not create duplicate comments', async () => {
    github.issues.getComments.mockReturnValueOnce(Promise.resolve({ data: [{
      body: '## I am an example title'
    }] }))

    await app.receive(event)
    expect(github.issues.createComment).not.toHaveBeenCalled()
  })

  it('creates many (5) comments', async () => {
    github.pullRequests.get.mockReturnValueOnce(loadDiff('many'))
    await app.receive(event)
    expect(github.issues.createComment).toHaveBeenCalledTimes(5)
  })

  it('ignores changes to the config file', async () => {
    github.pullRequests.get.mockReturnValueOnce(loadDiff('config'))
    await app.receive(event)
    expect(github.issues.createComment).not.toHaveBeenCalled()
  })

  it('ignores changes to the bin directory', async () => {
    github.pullRequests.get.mockReturnValueOnce(loadDiff('bin'))
    github.repos.getContent.mockReturnValueOnce(loadConfig('excludeBin'))
    await app.receive(event)
    expect(github.issues.createComment).not.toHaveBeenCalled()
  })

  it('works with a string as the keyword config', async () => {
    github.repos.getContent.mockReturnValueOnce(loadConfig('keywordsString'))
    await app.receive(event)
    expect(github.issues.createComment.mock.calls[0]).toMatchSnapshot()
  })

  it('creates a comment with a body line', async () => {
    github.pullRequests.get.mockReturnValueOnce(loadDiff('body'))
    await app.receive(event)
    expect(github.issues.createComment.mock.calls[0]).toMatchSnapshot()
  })

  it('deletes a comment on a removed TODO', async () => {
    github.pullRequests.get.mockReturnValueOnce(loadDiff('remove-todo'))
    github.issues.getComments.mockReturnValueOnce(Promise.resolve({ data: [{
      body: '## I am an example title\n\nHi!', id: 123
    }] }))
    await app.receive(event)
    expect(github.issues.editComment).toHaveBeenCalled()
    expect(github.issues.editComment).toHaveBeenCalledWith({
      comment_id: 123,
      owner: 'JasonEtco',
      repo: 'tests',
      number: pullRequestOpened.number,
      body: '<details><summary><strong>I am an example title</strong> âœ…</summary>\n\n## I am an example title\n\nHi!</details>'
    })
  })

  it('does nothing on a non-existant comment on a removed TODO', async () => {
    github.pullRequests.get.mockReturnValueOnce(loadDiff('remove-todo'))
    github.issues.getComments.mockReturnValueOnce(Promise.resolve({ data: [{
      body: '## I am not an example title'
    }] }))
    await app.receive(event)
    expect(github.issues.editComment).not.toHaveBeenCalled()
  })
})
