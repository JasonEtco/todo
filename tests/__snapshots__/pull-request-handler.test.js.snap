// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pull-request-handler comments on a pull request 1`] = `
Object {
  "body": "## I am an example title

https://github.com/JasonEtco/tests/blob/1409a3f61a27fe80a7170026fd73eaa2108a0108/.travis.yml#L10-L15

---

###### This comment was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 1409a3f61a27fe80a7170026fd73eaa2108a0108 in #32. cc @JasonEtco.",
}
`;

exports[`pull-request-handler comments on a pull request and mentions the assigned user 1`] = `
Object {
  "body": "## I am an example title

https://github.com/JasonEtco/tests/blob/1409a3f61a27fe80a7170026fd73eaa2108a0108/.travis.yml#L10-L15

---

###### This comment was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 1409a3f61a27fe80a7170026fd73eaa2108a0108 in #32. cc @matchai",
}
`;

exports[`pull-request-handler comments on a pull request and mentions the assigned users 1`] = `
Object {
  "body": "## I am an example title

https://github.com/JasonEtco/tests/blob/1409a3f61a27fe80a7170026fd73eaa2108a0108/.travis.yml#L10-L15

---

###### This comment was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 1409a3f61a27fe80a7170026fd73eaa2108a0108 in #32. cc @JasonEtco, @matchai and @defunkt",
}
`;

exports[`pull-request-handler creates a comment with a body line 1`] = `
Object {
  "body": "## I am an example title

Don&#x27;t forget about this body text!

---

https://github.com/JasonEtco/tests/blob/1409a3f61a27fe80a7170026fd73eaa2108a0108/.travis.yml#L10-L15

---

###### This comment was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 1409a3f61a27fe80a7170026fd73eaa2108a0108 in #32. cc @JasonEtco.",
}
`;

exports[`pull-request-handler creates many (5) comments 1`] = `
Object {
  "body": "## Title number 1

https://github.com/JasonEtco/tests/blob/1409a3f61a27fe80a7170026fd73eaa2108a0108/file.txt#L4

---

###### This comment was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 1409a3f61a27fe80a7170026fd73eaa2108a0108 in #32. cc @JasonEtco.",
}
`;

exports[`pull-request-handler creates many (5) comments 2`] = `
Object {
  "body": "## Title number 2

https://github.com/JasonEtco/tests/blob/1409a3f61a27fe80a7170026fd73eaa2108a0108/file.txt#L4

---

###### This comment was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 1409a3f61a27fe80a7170026fd73eaa2108a0108 in #32. cc @JasonEtco.",
}
`;

exports[`pull-request-handler creates many (5) comments 3`] = `
Object {
  "body": "## Title number 3

https://github.com/JasonEtco/tests/blob/1409a3f61a27fe80a7170026fd73eaa2108a0108/file.txt#L4

---

###### This comment was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 1409a3f61a27fe80a7170026fd73eaa2108a0108 in #32. cc @JasonEtco.",
}
`;

exports[`pull-request-handler creates many (5) comments 4`] = `
Object {
  "body": "## Title number 4

https://github.com/JasonEtco/tests/blob/1409a3f61a27fe80a7170026fd73eaa2108a0108/file.txt#L4

---

###### This comment was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 1409a3f61a27fe80a7170026fd73eaa2108a0108 in #32. cc @JasonEtco.",
}
`;

exports[`pull-request-handler creates many (5) comments 5`] = `
Object {
  "body": "## Title number 5

https://github.com/JasonEtco/tests/blob/1409a3f61a27fe80a7170026fd73eaa2108a0108/file.txt#L4

---

###### This comment was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 1409a3f61a27fe80a7170026fd73eaa2108a0108 in #32. cc @JasonEtco.",
}
`;

exports[`pull-request-handler works with a string as the keyword config 1`] = `
Object {
  "body": "## I am an example title

https://github.com/JasonEtco/tests/blob/1409a3f61a27fe80a7170026fd73eaa2108a0108/.travis.yml#L10-L15

---

###### This comment was generated by [todo](https://todo.jasonet.co) based on a \`pizza\` comment in 1409a3f61a27fe80a7170026fd73eaa2108a0108 in #32. cc @JasonEtco.",
}
`;
